server/api/src/rsDataObjChksum.cpp

```diff
@@ -1,97 +1,437 @@
+        for (auto&& row : irods::query{&rsComm, select_and_condition}) {
             try {
                 size_in_database = boost::lexical_cast<rodsLong_t>(row[0]);
-            } catch (const boost::bad_lexical_cast&) {
+            }
+            catch (const boost::bad_lexical_cast&) {
                 logger::api::error("_rsFileChksum: lexical_cast of [{}] for [{}] [{}] [{}] failed",
                     row[0],
-                    dataObjInfo->filePath,
-                    dataObjInfo->rescHier,
-                    dataObjInfo->objPath);
+                    dataObjInfo.filePath,
+                    dataObjInfo.rescHier,
+                    dataObjInfo.objPath);
                 return INVALID_LEXICAL_CAST;
             }
         }
+
         if (!size_in_database.has_value()) { return CAT_NO_ROWS_FOUND; }
         if (size_in_database.value() != size_in_vault) {
-            rodsLog(LOG_ERROR, "_rsFileChksum: file size mismatch. resource hierarchy [%s] vault path [%s] size [%ji] object path [%s] size [%ji]", dataObjInfo->rescHier, dataObjInfo->filePath,
-                    static_cast<intmax_t>(size_in_vault), dataObjInfo->objPath, static_cast<intmax_t>(size_in_database.value()));
+            rodsLog(LOG_ERROR, "_rsFileChksum: file size mismatch. resource hierarchy [%s] vault path [%s] size [%ji] object path [%s] size [%ji]",
+                    dataObjInfo.rescHier, dataObjInfo.filePath,
+                    static_cast<intmax_t>(size_in_vault), dataObjInfo.objPath, static_cast<intmax_t>(size_in_database.value()));
             return USER_FILE_SIZE_MISMATCH;
         }
+
         return 0;
+    } // verifyVaultSizeEqualsDatabaseSize
+
+    bool is_good_replica(const DataObjInfo& _replica) noexcept
+    {
+        return _replica.replStatus == GOOD_REPLICA;
+    } // is_stale_replica
+
+    bool is_member_of_bundle_resource(rodsLong_t _resource_id)
+    {
+        std::string resc_class;
+        const auto err = irods::get_resource_property<std::string>(_resource_id, irods::RESOURCE_CLASS, resc_class);
+        if (!err.ok()) {
+            logger::api::error(err.result());
+        }
+
+        return irods::RESOURCE_CLASS_BUNDLE == resc_class;
+    } // is_member_of_bundle_resource
+
+    std::vector<DataObjInfo*> get_good_replicas(DataObjInfo& _replicas)
+    {
+        std::vector<DataObjInfo*> good_replicas;
+        good_replicas.reserve(std::distance(begin(&_replicas), end(&_replicas)));
+
+        for (auto& r : &_replicas) {
+            if (is_good_replica(r) && !is_member_of_bundle_resource(r.rescId)) {
+                good_replicas.push_back(&r);
+            }
+        }
+
+        good_replicas.shrink_to_fit();
+
+        return good_replicas;
+    } // get_good_replicas
+
+    void report_number_of_replicas_skipped(const DataObjInfo* _replicas, rError_t& _error_stack)
+    {
+        std::size_t count = 0;
+
+        for (const auto& r : _replicas) {
+            if (!is_good_replica(r)) {
```

Comment:
the bundle resource needs to be considered here. just like in `get_good_replicas`.